### PR TITLE
Http server for OAuth

### PR DESCRIPTION
**Pull request by jliunyu**
1. Tested with the following change from librdkafka: https://github.com/jliunyu/librdkafka/commit/d7d1cbcf03c96c80a150a551853f282c7d05acf1

2. jliu@C02DM32YMD6T trivup % flake8 trivup/apps/OauthbearerOIDCApp.py                          
jliu@C02DM32YMD6T trivup % 

3. According to https://docs.confluent.io/platform/current/kafka/authorization.html: kafka.security.authorizer.AclAuthorizer was introduced at AK 2.4, it was called kafka.security.auth.SimpleAclAuthorizer before KA 2.4

### CODE REVIEW

File: trivup/apps/OauthbearerOIDCApp.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+        """
+        @param cluster     Current cluster.
+        @param conf        Configuration dict.
+        @param on          Node name to run on.
+
+        """
+        super(OauthbearerOIDCApp, self).__init__(cluster, conf=conf, on=on)
+
+    def start_cmd(self):
+        return 'python trivup/apps/OAuthbearerOIDCApp/HttpServer.py'
```
Comment (edenhill) [inline trivup/apps/OauthbearerOIDCApp.py]: This path is unlikely to work outside of the trivup source tree (e.g., when trivup is installed through pip).
You could instead do `python -m trivup.trivup.apps.OauthbearerOIDCApp.HttpServer` or something like that, but I think it is easier to just spin up a separate thread in Python instead.

File: trivup/apps/OauthbearerOIDCApp/HttpServer.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
+from http.server import BaseHTTPRequestHandler, HTTPServer
```
Comment (edenhill) [inline trivup/apps/OauthbearerOIDCApp/HttpServer.py]: Add copyright header

Comment (edenhill) [reply trivup/apps/OauthbearerOIDCApp/HttpServer.py]: Add python shebang

File: trivup/apps/OauthbearerOIDCApp/HttpServer.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+            self.send_response(200)
+            self.send_header('Content-type', 'text/html')
+            self.end_headers()
+            self.wfile.write(encoded_jwt.encode())
+        else:
+            self.send_error(404, 'URL is not valid: %s' % self.path)
+
+def main():
+    try:
+        port = 8000
```
Comment (edenhill) [inline trivup/apps/OauthbearerOIDCApp/HttpServer.py]: The port must be dynamically allocated to allow multiple instances. See the TCP allocator used by e.g. zookeeper.

File: trivup/apps/OauthbearerOIDCApp/HttpServer.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        curl \
+        -X POST \
+        --url localhost:8000/retrieve \
+        -H "accept: application/json" \
+        -H "authorization: Basic LW4gYWJjMTIzOlMzY3IzdCEK" \
+        -H "cache-control: no-cache" \
+        -H "content-type: application/x-www-form-urlencoded" \
+        -d "method=oidc,scope=test-scope"
+        '''
+        if self.path.endswith("/retrieve"):
```
Comment (edenhill) [inline trivup/apps/OauthbearerOIDCApp/HttpServer.py]: Avoid unneccessary indent for the common case by reversing this if-statement to:
```python
if not self.path.endswith...():
   self.send_error(404, ..)
   return

..
```

File: trivup/apps/OauthbearerOIDCApp/HttpServer.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+            content_length = int(self.headers['Content-Length'])
+            post_data = self.rfile.read(content_length)
+            
+            if not self.headers['Content-Type']: {
```
Comment (edenhill) [inline trivup/apps/OauthbearerOIDCApp/HttpServer.py]: Are these headers guaranteed to be in self.headers? If not you need to use ` self.headers.get(.., None)`.

Also use `if .. is None` rather than `if not`, since `if not` is a bit too smart.

File: trivup/apps/OauthbearerOIDCApp/HttpServer.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+                self.send_error(404, 'Content-Type field is required')
+            }
+            if not self.headers['cache-control']: {
```
Comment (edenhill) [inline trivup/apps/OauthbearerOIDCApp/HttpServer.py]: Why is this required?

File: trivup/apps/OauthbearerOIDCApp/HttpServer.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+                self.send_error(404, 'cache-control field is required')
+            }
+
+            if not self.headers['authorization']: {
```
Comment (edenhill) [inline trivup/apps/OauthbearerOIDCApp/HttpServer.py]: Shouldn't this be with a capital A?

File: trivup/apps/OauthbearerOIDCApp/HttpServer.py [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+                self.send_error(404, 'authorization field is required')
+            }
+
+            if not self.headers['Content-Length']: {
```
Comment (edenhill) [inline trivup/apps/OauthbearerOIDCApp/HttpServer.py]: This header is already retrieved at line 37, so this check comes too late.
Also see other comment on .get() use

File: trivup/apps/OauthbearerOIDCApp/HttpServer.py [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+                self.send_error(404, 'Content-Length field is required')
+            }
+
+            if not post_data:
```
Comment (edenhill) [inline trivup/apps/OauthbearerOIDCApp/HttpServer.py]: use 'is None'.
See https://stackoverflow.com/a/2710949/1821055

File: trivup/apps/OauthbearerOIDCApp/HttpServer.py [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+                self.send_error(404, 'method and scope fields are required in data')
+            
+            payloads = {"exp": datetime.datetime.utcnow() + \
+                               datetime.timedelta(seconds=30)}
+
+            encoded_jwt = jwt.encode(payloads, "secret", algorithm="HS256", 
+                                     headers={"kid": self.headers['authorization']})
+            self.send_response(200)
+            self.send_header('Content-type', 'text/html')
```
Comment (edenhill) [inline trivup/apps/OauthbearerOIDCApp/HttpServer.py]: Should it really be html?

File: trivup/clusters/KafkaCluster.py [hunk 11]
```diff
@@ -48,6 +49,7 @@
 #
 # See conf dict structure below.
 
+from trivup.apps.OauthbearerOIDCApp import OauthbearerOIDCApp
```
Comment (edenhill) [inline trivup/clusters/KafkaCluster.py]: Add it to the end of the imports, after SslApp.
I tend to order the imports in the priority, most critical first.

File: trivup/clusters/KafkaCluster.py [hunk 12]
```diff
@@ -442,6 +449,10 @@ def write_client_conf(self, path, additional_blob=None):
                         help='Existing Kafka source directory to use ' +
                         'instead of downloaded release. ' +
                         'Requires --version trunk.')
+    parser.add_argument('--oidc', dest='oidc',
+                        action='store_true',
+                        default=KafkaCluster.default_conf['oidc'],
+                        help='Start a HTTP server')
```
Comment (edenhill) [inline trivup/clusters/KafkaCluster.py]: Help text needs to say what it is for: "Enable Oauthbearer OIDC JWT server"

**Review by edenhill — CHANGES_REQUESTED** [review]
I think you should run the server as a thread rather than a sub-process, and move the implementation into the ..OIDCApp.py file.

File: trivup/apps/OauthbearerOIDCApp.py [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+        """
+        super(OauthbearerOIDCApp, self).__init__(cluster, conf=conf, on=on)
+        self.conf['port'] = trivup.TcpPortAllocator(self.cluster).next(
+            self, port_base=self.conf.get('port', None))
+
+    def start_cmd(self):
+        try:
+            server = HTTPServer(('', self.conf['port']), WebServerHandler)
+            print ("Web Server running on port %s" % self.conf['port'])
+            server.serve_forever()
```
Comment (jliunyu) [inline trivup/apps/OauthbearerOIDCApp.py]: Hi @edenhill , if we move the implementation to OauthbearerOIDCApp.py and start the server in this way, is extra thread to run the server still needed?

Comment (edenhill) [reply trivup/apps/OauthbearerOIDCApp.py]: I believe serve_forever() is a blocking call, so it would need to be run in a separate thread (or process).

File: trivup/apps/OauthbearerOIDCApp.py [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+                           otherwise.
+        @param on          Node name to run on.
+        """
+        super(OauthbearerOIDCApp, self).__init__(cluster, conf=conf, on=on)
+        self.conf['port'] = trivup.TcpPortAllocator(self.cluster).next(
+            self, port_base=self.conf.get('port', None))
+        self.conf['url'] = 'http://localhost:%d' % self.conf['port']
+
+    def start_cmd(self):
+        return "python -m trivup.apps.OauthbearerOIDCApp --port %s" % self.conf['port']
```
Comment (edenhill) [inline trivup/apps/OauthbearerOIDCApp.py]: Port should probably be an int (%d)?

File: trivup/apps/OauthbearerOIDCApp.py [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+    parser.add_argument('--port', type=int, dest='port',
+                        default=False,
+                        help='Port at which OauthbearerOIDCApp \
+                              should be bound')
+    args = parser.parse_args()
+    http_server = OauthbearerOIDCHttpServer()
+    http_server.enable_http_server(args.port)
+
+class OauthbearerOIDCApp (trivup.App):
+    """ Oauth/OIDC app, trigger an http server """
```
Comment (edenhill) [inline trivup/apps/OauthbearerOIDCApp.py]: "run a HTTP server"

File: trivup/apps/OauthbearerOIDCApp.py [hunk 16]
```diff
@@ -0,0 +1,149 @@
...
+    def enable_http_server(self, port):
+        server = HTTPServer(('', port), WebServerHandler)
+        thread = Thread(target=server.serve_forever, args=())
+        thread.start()
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(description= 'Trivup Oauthbearer OIDC \
+                                                   htter server')
+    parser.add_argument('--port', type=int, dest='port',
+                        default=False,
```
Comment (edenhill) [inline trivup/apps/OauthbearerOIDCApp.py]: Should be `required=true`

File: trivup/apps/OauthbearerOIDCApp.py [hunk 17]
```diff
@@ -0,0 +1,149 @@
...
+
+class OauthbearerOIDCHttpServer():
+    def enable_http_server(self, port):
+        server = HTTPServer(('', port), WebServerHandler)
+        thread = Thread(target=server.serve_forever, args=())
+        thread.start()
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(description= 'Trivup Oauthbearer OIDC \
+                                                   htter server')
```
Comment (edenhill) [inline trivup/apps/OauthbearerOIDCApp.py]: HTTP

File: trivup/apps/OauthbearerOIDCApp.py [hunk 18]
```diff
@@ -0,0 +1,149 @@
...
+        if self.headers.get('Content-Type', None) is None: {
+            self.send_error(404, 'Content-Type field is required')
+        }
+
+        if self.headers.get('Authorization', None) is None: {
+            self.send_error(404, 'Authorization field is required')
+        }
+
+        if post_data is None:
+            self.send_error(404,
```
Comment (edenhill) [inline trivup/apps/OauthbearerOIDCApp.py]: shouldnt this return?

File: trivup/apps/OauthbearerOIDCApp.py [hunk 19]
```diff
@@ -0,0 +1,149 @@
...
+        }
+
+        content_length = int(self.headers['Content-Length'])
+        post_data = self.rfile.read(content_length)
+
+        if self.headers.get('Content-Type', None) is None: {
+            self.send_error(404, 'Content-Type field is required')
+        }
+
+        if self.headers.get('Authorization', None) is None: {
```
Comment (edenhill) [inline trivup/apps/OauthbearerOIDCApp.py]: the braces are not needed

File: trivup/apps/OauthbearerOIDCApp.py [hunk 20]
```diff
@@ -0,0 +1,149 @@
...
+        if post_data is None:
+            self.send_error(404,
+                            'method and scope fields are required in data')
+
+        payloads = {"exp": datetime.datetime.utcnow() + \
+                           datetime.timedelta(seconds=30)}
+
+        encoded_jwt = jwt.encode(payloads, "secret", algorithm="HS256",
+                                 headers={"kid":
+                                          self.headers['authorization']})
```
Comment (edenhill) [inline trivup/apps/OauthbearerOIDCApp.py]: I believe dict keys are case-sensitive, and it is using Authorization above.

File: trivup/apps/OauthbearerOIDCApp.py [hunk 21]
```diff
@@ -0,0 +1,149 @@
...
+        self.wfile.write(message.encode())
+        print (message)
+
+    def do_POST(self):
+        """
+        Example usage:
+        curl \
+        -X POST \
+        --url localhost:8000/retrieve \
+        -H "accept: application/json" \
```
Comment (edenhill) [inline trivup/apps/OauthbearerOIDCApp.py]: Use upper-case, Accept. For all of these.

File: trivup/apps/OauthbearerOIDCApp.py [hunk 22]
```diff
@@ -0,0 +1,149 @@
...
+    def do_GET(self):
+        self.send_response(200)
+        self.send_header('Content-type', 'text')
+        self.end_headers()
+        message = "HTTP server for OAuth\n"
+        message += "Example for token retrieval:\n"
+        message += 'curl \
+        -X POST \
+        --url localhost:8000/retrieve \
+        -H "accept: application/json" \
```
Comment (edenhill) [inline trivup/apps/OauthbearerOIDCApp.py]: Use upper-case, Accept. For all of these.

File: trivup/apps/OauthbearerOIDCApp.py [hunk 23]
```diff
@@ -0,0 +1,149 @@
...
+                                 headers={"kid":
+                                          self.headers['authorization']})
+        self.send_response(200)
+        self.send_header('Content-type', 'text')
+        self.end_headers()
+        messages = {"access_token": encoded_jwt}
+        self.wfile.write(json.dumps(messages, indent = 4).encode())
+
+class OauthbearerOIDCHttpServer():
+    def enable_http_server(self, port):
```
Comment (edenhill) [inline trivup/apps/OauthbearerOIDCApp.py]: run_http_server

File: trivup/apps/OauthbearerOIDCApp.py [hunk 24]
```diff
@@ -0,0 +1,149 @@
...
+        server = HTTPServer(('', port), WebServerHandler)
```
Comment (edenhill) [inline trivup/apps/OauthbearerOIDCApp.py]: For security it should bind to localhost instead of any.

File: trivup/apps/OauthbearerOIDCApp.py [hunk 25]
```diff
@@ -0,0 +1,149 @@
...
+        thread = Thread(target=server.serve_forever, args=())
+        thread.start()
```
Comment (edenhill) [inline trivup/apps/OauthbearerOIDCApp.py]: With this thread in the background, what is keeping this process alive, isn't it reaching the end of __main__ after this and exiting?
I'm thinking you should perhaps run serve_forever() here instead of using a thread, but I'm not sure.

**Review by edenhill — CHANGES_REQUESTED** [review]
Good stuff!
Please run flake8 on the new file to fix any style issues.

Doesn't it also need to have an endpoint for the broker to query?

**Comment by edenhill** [comment]
Also make sure that you prepare for negative testing; an invalid token for instance, or something that can't be parsed, or can't be verified by the broker, etc.
E.g., "/retrieve/badformat" gives badly formatted jwt, "/retrieve/unverifiable" gives a token that can't be verified, etc.

**Comment by edenhill** [comment]
I think we'll want to start developing the librdkafka tests using this before we merge this, since it typically leads to some changes being made.

**Comment by jliunyu** [comment]
Hi @kirktrue, can you please also help to review this one?

File: trivup/apps/OauthbearerOIDCApp.py [hunk 26]
```diff
@@ -0,0 +1,221 @@
...
+    def generate_valid_token_for_client(self):
+        """
+        Example usage:
+        curl \
+        -X POST \
+        --url localhost:8000/retrieve \
+        -H "Accept: application/json" \
+        -H "Authorization: Basic LW4gYWJjMTIzOlMzY3IzdCEK" \
+        -H "Cache-Control: no-cache" \
+        -d "method=oidc,scope=test-scope"
```
Comment (kirktrue) [inline trivup/apps/OauthbearerOIDCApp.py]: I'm not familiar with the `method=oidc` parameter. According to my understanding of the spec, it should be `grant_type=client_credentials`. But this may be a different form I haven't seen.

File: trivup/apps/OauthbearerOIDCApp.py [hunk 27]
```diff
@@ -0,0 +1,221 @@
...
+        """
+        if self.headers.get('Content-Length', None) is None:
+            self.send_error(404, 'Content-Length field is required')
```
Comment (kirktrue) [inline trivup/apps/OauthbearerOIDCApp.py]: One could pedantically argue this should be a `400` status code, but since it's for testing, it probably doesn't matter.

The same argument applies for the other places that `404` is being returned.

File: trivup/apps/OauthbearerOIDCApp.py [hunk 28]
```diff
@@ -0,0 +1,221 @@
...
+            return
+
+        content_length = int(self.headers['Content-Length'])
+        post_data = self.rfile.read(content_length)
+
+        if self.headers.get('Cache-Control', None) != "no-cache":
```
Comment (kirktrue) [inline trivup/apps/OauthbearerOIDCApp.py]: I don't know if this `Cache-Control` header being set to `no-cache` is strictly a requirement 🤷‍♂️

File: trivup/apps/OauthbearerOIDCApp.py [hunk 29]
```diff
@@ -0,0 +1,221 @@
...
+            self.send_error(404, 'Authorization field is required')
+            return
+
+        if self.headers.get('Accept', None) != "application/json":
+            self.send_error(404, 'Accept field should be "application/json"')
+            return
+
+        if post_data is None:
+            self.send_error(404,
+                            'method and scope fields are required in data')
```
Comment (kirktrue) [inline trivup/apps/OauthbearerOIDCApp.py]: Per the above, for my own edification, can you point me to where the `method` parameter is described in the spec or on the Internet? My Google/DDG-fu isn't finding it. It would be helpful so I can update the Java client.

File: trivup/apps/OauthbearerOIDCApp.py [hunk 30]
```diff
@@ -0,0 +1,221 @@
...
+            return
+
+        new_key = RSA.generate(2048, e=65537)
+        public_key = new_key.publickey().exportKey("PEM")
+        private_key = new_key.exportKey("PEM")
+
+        public_keys.append(public_key)
+        payloads = {"exp": datetime.datetime.utcnow() +
```
Comment (kirktrue) [inline trivup/apps/OauthbearerOIDCApp.py]: The broker will require the `iat` and `sub` claims to appear in the JWT, so those should be given some sort of dummy value here.

File: trivup/apps/OauthbearerOIDCApp.py [hunk 31]
```diff
@@ -0,0 +1,221 @@
...
+                    datetime.timedelta(seconds=300)}
+
+        encoded_jwt = jwt.encode(payloads, private_key, algorithm="HS256",
+                                 headers={"kid":
```
Comment (kirktrue) [inline trivup/apps/OauthbearerOIDCApp.py]: The `kid` JWT header is the ID of the key, not the base-64-encoded client ID and secret.

You'll run into issues with the broker's ability to verify this key ID if the broker doesn't have access to the same set of keys. The broker will use this key ID to map to a public key, so the broker will need to somehow have access to that, likely through another endpoint exposed in this file. Perhaps that's what `response_to_broker` is intended to do?

File: trivup/apps/OauthbearerOIDCApp.py [hunk 32]
```diff
@@ -0,0 +1,221 @@
...
+                                          self.headers['Authorization']})
+        self.send_response(200)
+        self.send_header('Content-type', 'text')
+        self.end_headers()
+        messages = {"access_token": encoded_jwt}
+        self.wfile.write(json.dumps(messages, indent=4).encode())
+
+    def response_to_broker(self):
```
Comment (kirktrue) [inline trivup/apps/OauthbearerOIDCApp.py]: It looks like this is intended to be the response to the broker when it requests the JWKS. Is that correct? If so, the format needs to be changed.

Here's an example:

> https://login.microsoftonline.com/e162ef0e-e1d2-485e-888a-4972b54a45f3/discovery/v2.0/keys

**Review by kirktrue — COMMENTED** [review]
Take a look at the keys endpoint to determine how to format the OAuth keys response for JWKS.

File: trivup/apps/OauthbearerOIDCApp.py (hunk 26 above)
Comment (jliunyu) [reply trivup/apps/OauthbearerOIDCApp.py]: Hi @kirktrue， thanks for pointing out this, we used method=oidc to turn on the feature, but to token provider or broker, we should use grant_type=client_credentials, I will update this.

File: trivup/apps/OauthbearerOIDCApp.py (hunk 29 above)
Comment (jliunyu) [reply trivup/apps/OauthbearerOIDCApp.py]: Hi Kirk, never mind for this one, I will update it to grant_type=client_credentials

File: trivup/apps/KafkaBrokerApp.py [hunk 33]
```diff
@@ -219,9 +220,10 @@ def sort_listener(a):
                  self.conf['advertised.listeners'])
 
         if not self.kraft:
-            # SimpleAclAuthorizer doesn't work with KRaft, for some reason.
-            # Will investigate later.
-            conf_blob.append('authorizer.class.name=kafka.security.auth.SimpleAclAuthorizer')   # noqa: E501
+            if self.version < [2, 4, 0]:
+                conf_blob.append('authorizer.class.name=kafka.security.auth.SimpleAclAuthorizer')   # noqa: E501
```
Comment (edenhill) [inline trivup/apps/KafkaBrokerApp.py]: This is fixed on master, you can remove it here.

File: trivup/apps/OauthbearerOIDCApp.py [hunk 34]
```diff
@@ -0,0 +1,229 @@
...
+
+from jwcrypto import jwk
+import python_jwt as jwt
+from datetime import datetime, timedelta
+
+import json
+import argparse
+import requests
+
+public_keys = []
```
Comment (edenhill) [inline trivup/apps/OauthbearerOIDCApp.py]: Why is a global variable required?

File: trivup/apps/OauthbearerOIDCApp.py [hunk 35]
```diff
@@ -0,0 +1,229 @@
...
+    def do_GET(self):
+        if not self.path.endswith("/keys"):
+            self.send_response(200)
+            self.send_header('Content-type', 'application/json')
+            self.end_headers()
+            message = "HTTP server for OAuth\n"
+            message += "Example for token retrieval:\n"
+            message += 'curl \
+            -X POST \
+            --url localhost:8000/retrieve \
```
Comment (edenhill) [inline trivup/apps/OauthbearerOIDCApp.py]: Don't use fixed ports. Use the TCP port allocator instead.

File: trivup/apps/OauthbearerOIDCApp.py [hunk 36]
```diff
@@ -0,0 +1,229 @@
...
+            -H "Accept: application/json" \
+            -H "Content-type: application/x-www-form-urlencoded" \
+            -H "Authorization: Basic LW4gYWJjMTIzOlMzY3IzdCEK" \
```
Comment (edenhill) [inline trivup/apps/OauthbearerOIDCApp.py]: What's this base64(?) based on?

File: trivup/apps/OauthbearerOIDCApp.py (hunk 27 above)
Comment (edenhill) [reply trivup/apps/OauthbearerOIDCApp.py]: Yeah, I agree, 404 is Not found, so this should be 400.

File: trivup/apps/OauthbearerOIDCApp.py (hunk 24 above)
Comment (edenhill) [reply trivup/apps/OauthbearerOIDCApp.py]: Does not seem to be fixed?

**Review by edenhill — CHANGES_REQUESTED** [review]

File: trivup/apps/OauthbearerOIDCApp.py (hunk 35 above)
Comment (jliunyu) [reply trivup/apps/OauthbearerOIDCApp.py]: Hi @edenhill , thanks for reviewing.

This is just a curl command example that will return to users if they use `get` method by mistake. The real logic is using TCP port allocator: https://github.com/jliunyu/trivup/blob/oauth/trivup/apps/OauthbearerOIDCApp.py#L205

File: trivup/apps/OauthbearerOIDCApp.py (hunk 34 above)
Comment (jliunyu) [reply trivup/apps/OauthbearerOIDCApp.py]: I synced up with @kirktrue about how does broker sync up with the token provider before and pasted the answer from him below. From the OauthbearerOIDCApp, the public key was generated when starting it, I saved them to a global variable so if a scheduled sync up from broker happened, we can keep providing the public key without regenerating it.

`The relevant parts of the JWT are cached in the broker, yes. The only time the broker communicates with the OAuth/OIDC provider is at a) startup, b) during a scheduled refresh, and c) when it finds a key ID that isn't cached.
`

File: trivup/apps/OauthbearerOIDCApp.py [hunk 37]
```diff
@@ -0,0 +1,231 @@
...
+                             timedelta(seconds=lifetime), other_headers=header)
+    if not valid:
+        token += "invalid"
+
+    token_map = {"access_token": "%s" % token}
+    return token_map
+
+
+def generate_public_key():
+    global key
```
Comment (edenhill) [inline trivup/apps/OauthbearerOIDCApp.py]: I don't see the reason why this needs to be a global variable, it only seems to be used here in this method (and if it were a global it should have far more unique name).

It is vey seldom a good idea to use global variables.

trivup allows multiple Clusters to be instantiated separately, and each Cluster is isolated. And at the Cluster level, state should only be shared if absolutely needed, and in that case through querying the other Apps in that Cluster.
So, for this thing here we should keep all these keys on the WebServerHandler class instance object.

File: trivup/apps/OauthbearerOIDCApp.py [hunk 38]
```diff
@@ -0,0 +1,231 @@
...
+
+    key = jwk.JWK.generate(kty='RSA', size=2048, alg='RS256',
+                           use='sig', kid="abcdefg")
+
+    public_key = key.export_public()
+    public_keys.append(json.loads(public_key))
```
Comment (edenhill) [inline trivup/apps/OauthbearerOIDCApp.py]: Recommend just returning the key here and letting the caller append it to whatever it wants.

Comment (edenhill) [reply trivup/apps/OauthbearerOIDCApp.py]: Doesn't it need to store the private key too?

File: trivup/apps/OauthbearerOIDCApp.py (hunk 35 above)
Comment (edenhill) [reply trivup/apps/OauthbearerOIDCApp.py]: Even if it is example it should provide the right port, otherwise the example wont work.
At least change the fixed `:8000` to `:PORT`

File: trivup/apps/OauthbearerOIDCApp.py [hunk 39]
```diff
@@ -0,0 +1,231 @@
...
+            -H "Accept: application/json" \
+            -H "Content-type: application/x-www-form-urlencoded" \
+            -H "Authorization: Basic LW4gYWJjMTIzOlMzY3IzdCEK \
+                (base64 string generated from CLIENT_ID:CLIENT_SECRET)" \
+            -d "grant_type=client_credentials,scope=test-scope"'
+            self.wfile.write(message.encode())
+            return
+
+        self.send_response(200)
+        self.send_header('Content-type', 'application/json')
```
Comment (edenhill) [inline trivup/apps/OauthbearerOIDCApp.py]: nit: `Content-Type`

File: trivup/apps/OauthbearerOIDCApp.py [hunk 40]
```diff
@@ -0,0 +1,231 @@
...
+        -H "Content-type: application/x-www-form-urlencoded" \
+        -H "Authorization: Basic LW4gYWJjMTIzOlMzY3IzdCEK \
+            (base64 string generated from CLIENT_ID:CLIENT_SECRET)"
+        -d "grant_type=client_credentials,scope=test-scope"
+        """
+        if self.headers.get('Content-Length', None) is None:
+            self.send_error(400, 'Content-Length field is required')
+            return
+
+        if self.headers.get('Content-type', None) is None:
```
Comment (edenhill) [inline trivup/apps/OauthbearerOIDCApp.py]: `Content-Type` (also update the examples and other places).

File: trivup/apps/OauthbearerOIDCApp.py [hunk 41]
```diff
@@ -0,0 +1,231 @@
...
+
+import json
+import argparse
+import requests
+
+public_keys = []
+
+
+def generate_token(lifetime, valid=True):
+    private_pem = key.export_to_pem(private_key=True, password=None)
```
Comment (edenhill) [inline trivup/apps/OauthbearerOIDCApp.py]: where is `key` coming from?

File: trivup/apps/OauthbearerOIDCApp.py [hunk 42]
```diff
@@ -0,0 +1,231 @@
...
+    def run_http_server(self, port):
+        server = HTTPServer(('localhost', port), WebServerHandler)
+        server.serve_forever()
+
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(description='Trivup Oauthbearer OIDC \
+                                                  HTTP server')
+    parser.add_argument('--port', type=int, dest='port',
+                        default=False, required=True,
```
Comment (edenhill) [inline trivup/apps/OauthbearerOIDCApp.py]: default should hold the default value (and this is an int, not a bool), but since the argument is required there should be no default, just remove the param.

File: trivup/apps/OauthbearerOIDCApp.py (hunk 37 above)
Comment (edenhill) [reply trivup/apps/OauthbearerOIDCApp.py]: Any reason these methods can't be methods on the WebServerHandler class? That's the only context they're used in.

**Review by edenhill — CHANGES_REQUESTED** [review]

File: trivup/apps/OauthbearerOIDCApp.py (hunk 41 above)
Comment (jliunyu) [reply trivup/apps/OauthbearerOIDCApp.py]: The `key` is from the global variable: https://github.com/jliunyu/trivup/blob/oauth/trivup/apps/OauthbearerOIDCApp.py#L68

File: trivup/apps/OauthbearerOIDCApp.py [hunk 43]
```diff
@@ -0,0 +1,252 @@
...
+
+import json
+import argparse
+import requests
+
+
+class WebServerHandler(BaseHTTPRequestHandler):
+    def __init__(self):
+        self._key = None
+        self._public_keys = []
```
Comment (edenhill) [inline trivup/apps/OauthbearerOIDCApp.py:45]: I believe this class is instantiated per request, so these public keys would be created for each request and not shared.
I think this stuff should be on the OauthbearerOIDCHttpServer class, no?

**Review by edenhill — CHANGES_REQUESTED** [review]

File: trivup/apps/OauthbearerOIDCApp.py [hunk 44]
```diff
@@ -46,7 +46,26 @@ def __init__(self):
         self._mutex = Lock()
 
     def __call__(self, *args, **kwargs):
-        """ Handle a request """
+        """
+        As it turns out, http.server.BaseHTTPRequestHandler
```
Comment (edenhill) [inline trivup/apps/OauthbearerOIDCApp.py:50]: 💯

**Review by edenhill — APPROVED** [review]
LGTM!

**Comment by edenhill** [comment]
Thank you!
